Verify flake: verify-generated-protobuf.sh

```
+++ [0422 02:01:46] Creating data container
Error response from daemon: open /var/lib/docker/containers/39f63297738b06376d88f1295c428572710aa560223e3c79d9bcbef891e5da07/config.json: no such file or directory
!!! Error in /go/src/k8s.io/kubernetes/hack/../build/common.sh:595
  '"${docker_cmd[@]}" "$@"' exited with status 1
Call stack:
  1: /go/src/k8s.io/kubernetes/hack/../build/common.sh:595 kube::build::run_build_command(...)
  2: /go/src/k8s.io/kubernetes/hack/update-generated-protobuf.sh:57 main(...)
Exiting with status 1
!!! Error in /go/src/k8s.io/kubernetes/hack/after-build/verify-generated-protobuf.sh:44
  '"${KUBE_ROOT}/hack/update-generated-protobuf.sh"' exited with status 1
Call stack:
  1: /go/src/k8s.io/kubernetes/hack/after-build/verify-generated-protobuf.sh:44 main(...)
Exiting with status 1
!!! Error in ./hack/../hack/verify-generated-protobuf.sh:26
  '"${KUBE_ROOT}/hack/after-build/verify-generated-protobuf.sh" "$@"' exited with status 1
Call stack:
  1: ./hack/../hack/verify-generated-protobuf.sh:26 main(...)
Exiting with status 1
FAILED   ./hack/../hack/verify-generated-protobuf.sh	3s
```
https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/24291/kubernetes-pull-test-unit-integration/22388/

cc @kubernetes/sig-testing @wojtek-t

Ugh, it looks like builds are not properly isolated from each other.

Both kubernetes-pull-test-unit-integration/23398 and kubernetes-pull-test-unit-integration/23399 started at about the same time on Friday morning.

From 23398's build log:
```
Verifying ./hack/../hack/verify-generated-protobuf.sh
/go/src/k8s.io/kubernetes/hack/after-build/verify-generated-protobuf.sh: line 42: sudo: command not found
+++ [0429 03:26:14] Verifying Prerequisites....
+++ [0429 03:26:14] Building Docker image kube-build:build-91dbedbcaf.
+++ [0429 03:26:15] Running build command....
Go version: go version go1.6 linux/amd64
+++ [0429 03:26:16] Building go targets for linux/amd64:
    cmd/libs/go2idl/go-to-protobuf
    cmd/libs/go2idl/go-to-protobuf/protoc-gen-gogo
!!! Error in /go/src/k8s.io/kubernetes/hack/../build/common.sh:595
  '"${docker_cmd[@]}" "$@"' exited with status 137
Call stack:
  1: /go/src/k8s.io/kubernetes/hack/../build/common.sh:595 kube::build::run_build_command(...)
  2: /go/src/k8s.io/kubernetes/hack/update-generated-protobuf.sh:58 main(...)
Exiting with status 1
!!! Error in /go/src/k8s.io/kubernetes/hack/after-build/verify-generated-protobuf.sh:44
  '"${KUBE_ROOT}/hack/update-generated-protobuf.sh"' exited with status 1
Call stack:
  1: /go/src/k8s.io/kubernetes/hack/after-build/verify-generated-protobuf.sh:44 main(...)
Exiting with status 1
!!! Error in ./hack/../hack/verify-generated-protobuf.sh:26
  '"${KUBE_ROOT}/hack/after-build/verify-generated-protobuf.sh" "$@"' exited with status 1
Call stack:
  1: ./hack/../hack/verify-generated-protobuf.sh:26 main(...)
Exiting with status 1
FAILED   ./hack/../hack/verify-generated-protobuf.sh	7s
```

From 23399's build log:
```
Verifying ./hack/../hack/verify-generated-protobuf.sh
/go/src/k8s.io/kubernetes/hack/after-build/verify-generated-protobuf.sh: line 42: sudo: command not found
+++ [0429 03:26:20] Verifying Prerequisites....
+++ [0429 03:26:20] Building Docker image kube-build:build-91dbedbcaf.
+++ [0429 03:26:20] Running build command....
Go version: go version go1.6 linux/amd64
+++ [0429 03:26:22] Building go targets for linux/amd64:
    cmd/libs/go2idl/go-to-protobuf
    cmd/libs/go2idl/go-to-protobuf/protoc-gen-gogo
+++ [0429 03:26:27] Placing binaries
type checker error: github.com/opencontainers/runc/libcontainer/cgroups/fs/cpuacct.go:22:25: GetClockTicks not declared by package system
type checking encountered some errors in "github.com/opencontainers/runc/libcontainer/cgroups/fs", but ignoring.
2016/04/29 03:26:29 Processing package "k8s.io.kubernetes.pkg.util.intstr", disk location "/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/intstr"
...
SUCCESS  ./hack/../hack/verify-generated-protobuf.sh	19s
```

Note how the docker build for each started within seconds of each other, and they both used the same image name.

--- a/build/common.sh
+++ b/build/common.sh
@@ -154,7 +154,7 @@ function kube::build::verify_prereqs() {
   fi
   kube::build::ensure_docker_daemon_connectivity || return 1
 
-  KUBE_ROOT_HASH=$(kube::build::short_hash "$KUBE_ROOT")
+  KUBE_ROOT_HASH=$(kube::build::short_hash "${HOSTNAME:-}:${KUBE_ROOT}")
   KUBE_BUILD_IMAGE_TAG="build-${KUBE_ROOT_HASH}"
   KUBE_BUILD_IMAGE="${KUBE_BUILD_IMAGE_REPO}:${KUBE_BUILD_IMAGE_TAG}"
   KUBE_BUILD_CONTAINER_NAME="kube-build-${KUBE_ROOT_HASH}"

--- a/hack/after-build/verify-generated-protobuf.sh
+++ b/hack/after-build/verify-generated-protobuf.sh
@@ -36,10 +36,12 @@ for APIROOT in ${APIROOTS}; do
   cp -a "${KUBE_ROOT}/${APIROOT}" "${_tmp}/${APIROOT}"
 done
 
-# We would like to use "sudo" when running on Travis and
-# not use "sudo" when running on Jenkins.
-SUDO="sudo"
-sudo -h > /dev/null || SUDO=""
+# If not running as root, we need to use sudo to restore the original generated
+# protobuf files.
+SUDO=""
+if [[ "$(id -u)" != '0' ]]; then
+  SUDO="sudo"
+fi
 
 "${KUBE_ROOT}/hack/update-generated-protobuf.sh"
 for APIROOT in ${APIROOTS}; do

--- a/hack/update-generated-protobuf.sh
+++ b/hack/update-generated-protobuf.sh
@@ -31,7 +31,7 @@ function prereqs() {
   fi
   kube::build::ensure_docker_daemon_connectivity || return 1
 
-  KUBE_ROOT_HASH=$(kube::build::short_hash "$KUBE_ROOT/go-to-protobuf")
+  KUBE_ROOT_HASH=$(kube::build::short_hash "${HOSTNAME:-}:${REPO_DIR:-${KUBE_ROOT}}/go-to-protobuf")
   KUBE_BUILD_IMAGE_TAG="build-${KUBE_ROOT_HASH}"
   KUBE_BUILD_IMAGE="${KUBE_BUILD_IMAGE_REPO}:${KUBE_BUILD_IMAGE_TAG}"
   KUBE_BUILD_CONTAINER_NAME="kube-build-${KUBE_ROOT_HASH}"

